src/libmeasurement_kit/ooni/dpi_fragment.cpp

```diff
@@ -0,0 +1,373 @@
+                    logger->info("$$$$$$ found plaintext hostname; hacking planet $$$$$$$$$");
+                    n = write(sockfd, nwrite->c_str(), found+3);
+                    if (n > 0) {
+                        nwrite->erase(0, n);
+                    }
+                    sleep(0.5);
+                }
+            }
+
+            n = write(sockfd, nwrite->c_str(), nwrite->length());
+            if (n > 0) {
+                nwrite->erase(0, n);
+            }
+        }
+
+        if (FD_ISSET(sockfd, &rfds)) {
+            logger->info("READABLE\n");
+            while(1) {
+                n = read(sockfd, &readbuf, sizeof(readbuf));
+                logger->info("read from sockfd: %d", n);
+                if (read > 0) {
+                    size_t cur_size = nread->length();
+                    nread->resize(cur_size + n);
+                    std::copy(readbuf, readbuf + n, nread->begin() + cur_size);
+                }
+                if (n == 0) {
+                    stop_reading = true;
+                    break;
+                }
+                if (static_cast<size_t>(n) != sizeof(readbuf)) {
+                    break;
+                }
+                if(n < 0) {
+                    break;
+                }
+            }
+        }
+
+        if (do_ssl) {
+            ssl_filter.update();
+        }
+        if (awrite->size() == 0 && nwrite->size() == 0) {
+            stop_writing = true;
+        }
+        if (stop_reading && stop_writing) {
+            break;
+        }
+
+    }
+
+    return *aread;
+}
+
+
+
+
+void dpi_fragment(Settings options, Callback<SharedPtr<report::Entry>> callback,
+                        SharedPtr<Reactor> reactor, SharedPtr<Logger> logger) {
+    reactor->call_in_thread(logger, [=]() {
+        logger->info("starting dpi_fragment");
+        SharedPtr<Entry> entry(new Entry);
+
+        std::string fragmented_https = fragmented_https_request(true, true, logger);
+        std::string unfragmented_https = fragmented_https_request(false, true, logger);
+        std::string fragmented_http = fragmented_https_request(true, false, logger);
+        std::string unfragmented_http = fragmented_https_request(false, false, logger);
+
+        logger->info("fragmented https response: %s", fragmented_https.c_str());
+        logger->info("unfragmented https response: %s", unfragmented_https.c_str());
+        logger->info("fragmented http response: %s", fragmented_http.c_str());
+        logger->info("unfragmented http response: %s", unfragmented_http.c_str());
+
+        logger->info("fragmented https response length: %d", fragmented_https.length());
+        logger->info("unfragmented https response length: %d", unfragmented_https.length());
+        logger->info("fragmented http response length: %d", fragmented_http.length());
+        logger->info("unfragmented http response length: %d", unfragmented_http.length());
+        callback(entry);
```

Comment:
I suggest to do

```C++
reactor->call_soon([entry = std::move(entry), callback = std::move(callback)]() {
    callback(entry);
});
```

so that the callback is called from the I/O thread loop, which is what usually happens.

Reply:
Done.